plugins/nvme/fu-nvme-device.c

```diff
@@ -0,0 +1,406 @@
+{
+	struct nvme_admin_cmd cmd = {
+		.opcode		= 0x10,
+		.cdw10		= (bpid << 31) | (action << 3) | slot,
+	};
+	return fu_nvme_device_submit_admin_passthru (self, &cmd, error);
+}
+
+static gboolean
+fu_nvme_device_fw_download (FuNvmeDevice *self,
+			    guint32 addr,
+			    const guint8 *data,
+			    guint32 data_sz,
+			    GError **error)
+{
+	struct nvme_admin_cmd cmd = {
+		.opcode		= 0x11,
+		.addr		= 0x0, /* memory address of data */
+		.data_len	= data_sz,
+		.cdw10		= (data_sz >> 2) - 1,
+		.cdw11		= addr >> 2,
+	};
+	memcpy (&cmd.addr, &data, sizeof (gpointer));
+	return fu_nvme_device_submit_admin_passthru (self, &cmd, error);
+}
+
+static void
+fu_nvme_device_parse_cns_maybe_dell (FuNvmeDevice *self, const guint8 *buf)
+{
+	g_autofree gchar *component_id = NULL;
+	g_autofree gchar *guid_efi = NULL;
+	g_autofree gchar *guid_id = NULL;
+
+	/* add extra component ID if set */
+	component_id = fu_nvme_device_get_string_safe (buf, 0xc36, 0xc3d);
+	if (component_id == NULL || strlen (component_id) < 4) {
+		g_debug ("invalid component ID, skipping");
+		return;
+	}
+	guid_id = g_strdup_printf ("STORAGE-DELL-%s", component_id);
+	fu_device_add_guid (FU_DEVICE (self), guid_id);
+
+	/* also add the EFI GUID */
+	guid_efi = fu_nvme_device_get_guid_safe (buf, 0x0c26);
+	if (guid_efi != NULL)
+		fu_device_add_guid (FU_DEVICE (self), guid_efi);
+}
+
+static gboolean
+fu_nvme_device_parse_cns (FuNvmeDevice *self, const guint8 *buf, gsize sz, GError **error)
+{
+	guint8 fawr;
+	guint8 nfws;
+	guint8 s1ro;
+	g_autofree gchar *mn = NULL;
+	g_autofree gchar *sn = NULL;
+	g_autofree gchar *sr = NULL;
+
+	/* wrong size */
+	if (sz != FU_NVME_ID_CTRL_SIZE) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "failed to parse blob, expected 0x%04x bytes",
+			     (guint) FU_NVME_ID_CTRL_SIZE);
+		return FALSE;
+	}
+
+	/* get sanitiezed string from CNS -- see the following doc for offsets:
+	 * NVM-Express-1_3c-2018.05.24-Ratified.pdf */
+	sn = fu_nvme_device_get_string_safe (buf, 4, 23);
+	if (sn != NULL)
+		fu_device_set_serial (FU_DEVICE (self), sn);
+	mn = fu_nvme_device_get_string_safe (buf, 24, 63);
+	if (mn != NULL)
+		fu_device_set_name (FU_DEVICE (self), mn);
+	sr = fu_nvme_device_get_string_safe (buf, 64, 71);
+	if (sr != NULL)
+		fu_device_set_version (FU_DEVICE (self), sr);
+
+	/* firmware slot information */
+	fawr = (buf[260] & 0x10) >> 4;
+	nfws = (buf[260] & 0x0e) >> 1;
+	s1ro = buf[260] & 0x01;
+	g_debug ("nr fw slots: %u, slot1 r/o: %u", nfws, s1ro);
+	if (!fawr)
+		fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_NEEDS_REBOOT);
+
+	/* Dell helpfully provide an EFI GUID we can use in the vendor offset,
+	 * but don't have a header or any magic we can use -- so check if the
+	 * component ID looks plausible and the GUID is "sane" */
+	fu_nvme_device_parse_cns_maybe_dell (self, buf);
+
+	/* fall back to the device description */
+	if (fu_device_get_guids (FU_DEVICE (self))->len == 0) {
```

Comment:
In your data set were you able to check if any other vendors 'vs' blobs set off the Dell creation of a Dell GUID?  With this approach I would just wonder if another vendor vs section has a set of bytes that could represent an EFI guid at that offset) (but doesn't actually) what happens.

Reply:
It takes more than the plausible GUID, it also takes a plausible component ID string too.

Reply:
OK, I guess that makes sense then for this.  It would only apply to "newer" disks (that only have one field), but it's safer.